pkg/kubelet/cm/dra/manager.go

```diff
@@ -383,19 +509,15 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 				return fmt.Errorf("NodeUnprepareResources failed for claim %s/%s: %s", reqClaim.Namespace, reqClaim.Name, result.Error)
 			}
 
-			// Delete last pod UID only if unprepare succeeds.
+			claimName := claimNamesMap[types.UID(claimUID)]
+
+			// Delete claim info from the cache only when unprepare succeeds.
 			// This ensures that the status manager doesn't enter termination status
 			// for the pod. This logic is implemented in
 			// m.PodMightNeedToUnprepareResources and claimInfo.hasPodReference.
-			claimInfo := claimInfos[types.UID(claimUID)]
-			claimInfo.deletePodReference(pod.UID)
-			m.cache.delete(claimInfo.ClaimName, pod.Namespace)
-		}
-
-		// Checkpoint to reduce redundant calls to NodeUnprepareResources after a kubelet restart.
-		err = m.cache.syncToCheckpoint()
-		if err != nil {
-			return fmt.Errorf("failed to checkpoint claimInfo state, err: %+v", err)
+			m.cache.Lock()
+			m.cache.delete(claimName, namespace)
```

Comment:
@klueska This would remove the claim from the cache when first UnprepareResources succeeds for it. If there are more than one plugins and second plugin fails UnprepareResources call, the call will never be repeated as the claim is not in the cache anymore.

Reply:
Yep, good catch. We only want to delete after the last unprepare call just before syncing to the checkpoint.

Reply:
Yep, good catch. This uses to just delete the pod reference and way changed during some iteration of the code. It should be changed back to only removing the pod reference, not the whole claimInfo.